### PR TITLE
Game logic

### PR DESCRIPTION
**Pull request by Yors-git**
In this branch we worked on the following:
- We separated the classes and methods in different files
- We worked on the logic to validate the inputs of the players, to make sure it's a number as requested and that its not repeated
- We worked on the logic to validate which player is the winner

### CODE REVIEW

**Review by jubaan — CHANGES_REQUESTED** [review]
# Status: Changes Required :recycle: 

## :wave: :wave: Hello @Yors-git  and @Ceejayski  . Thanks for submitting you project!

Things to notice:
:100: the game has a good looking board
:100: The players switch
:100: The game is playable with some bugs

There are some issues that need to be addressed before we can move forward. Please review the list below.

## Required changes :wrench: 
### [Mandatory] :red_circle: 

- [x] Configure the linters, the project doesn't have GitHub Actions nor local version of the linters.

- [x] The game let nameless players to play. This makes it hard to keep track of the player in the game.

- [x] The should not be any puts/print outside the ```bin/main``` file. The board uses puts.

- [x] use the ```!``` negation instead of using ```condition == false````. See inline comments.

- [x] use ```require_relative```

- [x] Don't disable Rubocop cops and fix the offenses that appear. The only permitted ones are related to the method/class/module length and complexity.
- Style/Documentation
- Style/GuardClause 
- Style/MixinUsage
- ... others

- [x] ```Include``` should be used only inside a class or a module.

- [x] The game is ending early
image

- [x] append the ```?``` question mark to the method names that return boolean values (```true```` or ```false```). Example: ```is_valid?```. This is best practices and improves readability, a programmer understands the output without looking at the code.

- [x] The game doesn't show the latest board status.
image

- [x] The game has difficulties identifying the winner
image


### [OPTIONAL]  :yellow_circle:  

- [x] Use a string block for the board instead of one line strings. This will only require one puts to print the whole board instead of 5 and will have the same result.

- [x] the ```cases``` variable is more a constant than a variable. It would be better to show the intent and ensure immutability with the ```.freeze``` method.

- [x] Use space between blocks to separate visually the elements. In some areas, the code looks compact. This will improve readability of the code.

### [Comments] :green_circle: 

- 

## Keep up the great work :muscle: and Happy Coding :smiley: :computer:!!

File: lib/game_logic.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  def display_board(board)
+    puts " #{board[0]} | #{board[1]} | #{board[2]} "
+    puts '-----------'
+    puts " #{board[3]} | #{board[4]} | #{board[5]} "
+    puts '-----------'
+    puts " #{board[6]} | #{board[7]} | #{board[8]} "
+  end
+
+  def valid_input(input, arr)
+    if (input.nil? || input.is_a?(Integer) == false || (1..9).include?(input) == false) || (arr[input - 1] == 'X' || arr[input - 1] == 'O')
```
Comment (jubaan) [inline lib/game_logic.rb:14]: This ```input.is_a?(Integer) == false``` is equal to ```!input.is_a?(Integer)``` the only addition is the ```!``` exclamation sign. this way you can remove the false to make the line shorter.

File: lib/game_logic.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      false
+    end
+  end
```
Comment (jubaan) [inline lib/game_logic.rb]: This ```input.is_a?(Integer) == false``` is equal to ```!input.is_a?(Integer)``` the only addition is the ```!``` exclamation sign. this way you can remove the false to make the line shorter.

**Review by jubaan — CHANGES_REQUESTED** [review]

File: lib/game_logic.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      case board_case
+      when 'XXX'
+        win = 'X'
+      when 'OOO'
+        win = 'O'
+      end
+    end
+    win
+  end
+end
```
Comment (maelfosso) [inline lib/game_logic.rb]: Remove `self`, please. No static methods.
Add attributes as describe up there.

File: bin/main [hunk 4]
```diff
@@ -1,53 +1,85 @@
...
+require_relative '../lib/players'
+
+puts 'Welcome to Tic-Tac-Toe'
+sleep(1)
+puts 'Before we start may we know your names'
+sleep(1)
 
 game_on = true
+
 until game_on == false
```
Comment (maelfosso) [inline bin/main]: **OPTIONAL** As indicated by the previous reviewer, is better to write `!game_one`

File: bin/main [hunk 5]
```diff
@@ -1,53 +1,85 @@
...
+  end
+  puts 'Player 2 please enter your name'
+  player2 = Player.new(gets.chomp)
+  until player2.valid_name?
+    puts 'Please enter a valid name'
+    player2.name = gets.chomp
+  end
+  puts
+  game_board = [1, 2, 3, 4, 5, 6, 7, 8, 9]
+  puts GameLogic.display_board(game_board)
```
Comment (maelfosso) [inline bin/main]: Set `game_board` as an attribute of `GameLogic` class and instantiate the `GameLogic` then use the instantiated object

File: bin/main [hunk 6]
```diff
@@ -1,53 +1,85 @@
...
+
+  count = 0
+
+  while count < 9
+    puts
+    puts "#{active_player.name} select the number (1-9) for your move"
+    puts
+    current_move = gets.chomp.to_i
+
+    if GameLogic.valid_input?(current_move, game_board) == false
```
Comment (maelfosso) [inline bin/main]: Same remark, after instantiating, call it like this
```rb
game.valid_input?(current_move)
```
The `game_board` will already be a GameLogic attribute

File: bin/main [hunk 7]
```diff
@@ -1,53 +1,85 @@
...
+      active_player = player1
+      mark = active_player.player1_mark
     end
-    # puts "#{winner} you won!"
-    puts "Play again? press \"Y\" or press any other key to quit"
-    input = gets.chomp.upcase
-    if input == "Y"
-        game_on = true
+
+    case GameLogic.winning_cases(game_board)
```
Comment (maelfosso) [inline bin/main]: Replace it by
```rb
game.winning_cases()
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Thank you for making the suggested changes.
You did a really good job so far, the game is working like a charm.

However, your OOP approach is not good.
1. Let's remind some things concerning the Tic-Tac-Toe Game

- The Game doesn't exists if there is no Board or Players. If we agree on that, it means that you must have a Game class with two/three attributes: Board, Players

- As the method, we have all the different actions with can be done on the Game. All the actions concerning puts/gets are not taken in consideration here. Because, at the end, the `lib` folder must be use as any other library in any kind of Ruby/Ruby on Rails projects.

2- Kindly read the comments left below to be guided on how to refactor your code

Once done, submit a new code review request
Thanks

Mael FOSSO

File: bin/main (hunk 4 above)
Comment (Yors-git) [reply bin/main]: Solved!

File: lib/game_logic.rb (hunk 3 above)
Comment (Yors-git) [reply lib/game_logic.rb]: Removed Self

File: bin/main (hunk 5 above)
Comment (Yors-git) [reply bin/main]: Removed!

File: bin/main (hunk 6 above)
Comment (Yors-git) [reply bin/main]: Solved

File: bin/main (hunk 7 above)
Comment (Yors-git) [reply bin/main]: solved!

**Review by kevcap — APPROVED** [review]
# [STATUS] PROJECT APPROVED 🍾

### I really like your project! Congrats team, you guys did an excellent job, you have my approval to go ahead and merge this pull request!



## As always, I wish you all the very best and see you both at the top🚀


👨🏻‍💻 **TSE: Kevin Alves**

  

📫 Have any question? Better call Kevin:

<a  href="mailto:kevinalves.js@gmail.com"> gmail</a> [Twitter Follow](https://twitter.com/intent/follow?screen_name=kevinnnmn) [](https://microverse-students.slack.com/team/UT86D3AP4)

**Pull request merged**